app/controllers/challenge_partnerships_controller.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+class ChallengePartnershipsController < ApplicationController
+  before_action :set_form, only: :create
+
+  def show
+    set_form
+  rescue TokenExpiredError
+    redirect_to link_expired_challenge_partnership_path
+  rescue AlreadyChallengedError
+    redirect_to already_challenged_challenge_partnership_path(school_name: @school_name)
+  end
+
+  def create
+    render :show and return unless @challenge_partnership_form.valid?
+
+    @challenge_partnership_form.challenge!
+    redirect_to success_challenge_partnership_path
+  end
+
+  def link_expired; end
+
+  def already_challenged
+    @school_name = params[:school_name]
+  end
+
+  def success; end
+
+private
+
+  def set_form
+    token = params[:token] || params.dig(:challenge_partnership_form, :token)
+    notification_email = PartnershipNotificationEmail.find_by(token: token)
+    raise ActionController::RoutingError, "Not Found" if notification_email.blank?
+    raise TokenExpiredError if notification_email.token_expired?
+
+    @partnership = notification_email.partnership
+    @school_name = @partnership.school.name
+    raise AlreadyChallengedError if @partnership.challenged?
```

Comment:
double checking, do school name and partnership need to be `@` variables? We will have them on the form too.

Reply:
Yup, we use them in some of the rescues, before we have the form